File: sdk/python/src/basanos_sdk/watcher.py
"""Watcher-side monitoring and challenge-descent helpers."""
from __future__ import annotations

import hashlib
from collections.abc import Sequence
from dataclasses import dataclass

from solders.instruction import Instruction
from solders.keypair import Keypair
from solders.pubkey import Pubkey

from . import consensus as c
from . import instructions as ix
from .transport import RpcClient


@dataclass(frozen=True)
class WatcherAccounts:
    dcg_program: bytes
    pt2s: bytes
    pt1s: bytes
    routes: bytes
    geometry: bytes
    registry: bytes
    pt2s_sha256: bytes | None = None

    @classmethod
    def create(cls, dcg_program: Pubkey | str | bytes | bytearray, *, pt2s: Pubkey | str | bytes | bytearray,
               pt1s: Pubkey | str | bytes | bytearray, routes: Pubkey | str | bytes | bytearray,
               geometry: Pubkey | str | bytes | bytearray, registry: Pubkey | str | bytes | bytearray,
               pt2s_sha256: bytes | None = None) -> WatcherAccounts:
        return cls(c.key_bytes(dcg_program), c.key_bytes(pt2s), c.key_bytes(pt1s), c.key_bytes(routes),
                   c.key_bytes(geometry), c.key_bytes(registry), bytes(pt2s_sha256) if pt2s_sha256 is not None else None)


@dataclass(frozen=True)
class DescentRound:
    descendants: tuple[bytes, ...]
    tree_root: bytes | None
    choice: int
    fixpoint: bool = False


@dataclass(frozen=True)
class ChallengeHandle:
    descriptor: bytes
    record: bytes
    nonce: int
    position: int
    instruction: Instruction
    response: bytes = b""


class Watcher:
    """Read document state and construct challenge/descent instructions."""

    def __init__(self, rpc: RpcClient, dcg_program: Pubkey | str | bytes | bytearray, accounts: WatcherAccounts) -> None:
        self.rpc = rpc
        self.accounts = accounts

    def book(self, descriptor: bytes, *, challenger: Pubkey | str | bytes | bytearray | None = None,
             nonce: int = 0, position_count: int | None = None) -> c.AddressBook:
        return c.addresses(self.accounts.dcg_program, descriptor=descriptor, pt2s=self.accounts.pt2s,
                           pt2s_sha256=self.accounts.pt2s_sha256, challenger=challenger,
                           nonce=nonce, registry=self.accounts.registry, position_count=position_count)

    def read_document(self, descriptor: bytes) -> c.Dcm2V5:
        book = self.book(descriptor)
        account = self.rpc.read_account(book.document[0])
        if account is None:
            raise LookupError("document account is absent")
        return c.Dcm2V5.decode(account.data)

    def positions(self, descriptor: bytes) -> list[bytes]:
        book = self.book(descriptor)
        account = self.rpc.read_account(book.positions[0])
        if account is None:
            raise LookupError("positions account is absent")
        _descriptor, roots = c.decode_dpr2(account.data, allow_partial=True)
        return roots

    def list_positions(self, descriptor: bytes) -> list[tuple[int, bytes]]:
        return list(enumerate(self.positions(descriptor)))

    def pick_sample(self, descriptor: bytes, position_count: int, *, nonce: int = 0) -> int:
        if position_count < 1 or not 0 <= nonce < 1 << 64:
            raise ValueError("invalid position count or nonce")
        value = hashlib.sha256(c.digest(descriptor) + c.uint(nonce, 8)).digest()
        return int.from_bytes(value[:8], "little") % position_count

    def open_position(self, descriptor: bytes, challenger: Pubkey | str | bytes | bytearray, position: int,
                      *, response_length: int, nonce: int) -> ChallengeHandle:
        who = c.key_bytes(challenger)
        book = self.book(descriptor, challenger=who, nonce=nonce)
        record = book.challenge[0]
        data = c.encode_challenge_position(descriptor, position, response_length, nonce)
        instruction = ix.challenge_position(self.accounts.dcg_program, record, who, book.document[0], book.positions[0],
                                            self.accounts.pt2s, self.accounts.routes, self.accounts.geometry,
                                            self.accounts.registry, self.accounts.pt1s, data)
        return ChallengeHandle(descriptor, record, nonce, position, instruction, book.response[0])

    def open_leaf(self, descriptor: bytes, challenger: Pubkey | str | bytes | bytearray, *, position: int,
                  segment: int, local: int, leaf: bytes, path: Sequence[bytes], spp1: bytes,
                  response_length: int, nonce: int) -> ChallengeHandle:
        who = c.key_bytes(challenger)
        book = self.book(descriptor, challenger=who, nonce=nonce)
        record = book.challenge[0]
        data = c.encode_challenge_leaf(descriptor, position, segment, local, leaf, path, spp1, response_length, nonce)
        instruction = ix.challenge_leaf(self.accounts.dcg_program, record, who, book.document[0], book.positions[0],
                                        self.accounts.pt2s, self.accounts.routes, self.accounts.geometry,
                                        self.accounts.registry, self.accounts.pt1s, data)
        return ChallengeHandle(descriptor, record, nonce, position, instruction, book.response[0])

    def reveal_position(self, handle: ChallengeHandle, executor: Pubkey | str | bytes | bytearray,
                        first: int, roots: Sequence[bytes]) -> Instruction:
        book = self.book(handle.descriptor)
        return ix.reveal_position(self.accounts.dcg_program, handle.record, executor, book.document[0], book.positions[0],
                                  self.accounts.pt2s, self.accounts.routes, self.accounts.geometry, first, roots)

    def select_segment(self, handle: ChallengeHandle, challenger: Pubkey | str | bytes | bytearray,
                       ordinal: int) -> Instruction:
        book = self.book(handle.descriptor)
        return ix.select_segment(self.accounts.dcg_program, handle.record, challenger, book.document[0],
                                 self.accounts.pt2s, self.accounts.routes, self.accounts.geometry, ordinal)

    def close_response(self, handle: ChallengeHandle, executor: Pubkey | str | bytes | bytearray) -> Instruction:
        response = handle.response or c.pda(self.accounts.dcg_program, c.RESPONSE_SEED, handle.record)[0]
        return ix.close_response(self.accounts.dcg_program, handle.record, response, executor)

    def settle(self, handle: ChallengeHandle, winner: Pubkey | str | bytes | bytearray,
               challenger: Pubkey | str | bytes | bytearray,
               executor: Pubkey | str | bytes | bytearray | None = None) -> Instruction:
        response = handle.response or c.pda(self.accounts.dcg_program, c.RESPONSE_SEED, handle.record)[0]
        executor = self.read_challenge(handle.record)["executor"] if executor is None else executor
        book = self.book(handle.descriptor)
        return ix.settle(self.accounts.dcg_program, handle.record, response, winner, executor,
                         book.document[0], challenger)

    def _fixpoint_accounts(self) -> list[tuple[bytes, bool, bool]]:
        return [(self.accounts.pt2s, False, False), (self.accounts.routes, False, False),
                (self.accounts.geometry, False, False), (self.accounts.registry, False, False),
                (self.accounts.pt1s, False, False)]

    def descent_instructions(self, handle: ChallengeHandle, executor: Pubkey | str | bytes | bytearray,
                             challenger: Pubkey | str | bytes | bytearray,
                             rounds: Sequence[DescentRound]) -> tuple[Instruction, ...]:
        book = self.book(handle.descriptor)
        result: list[Instruction] = []
        for round_ in rounds:
            fixpoint = self._fixpoint_accounts() if round_.fixpoint else None
            result.append(ix.reveal_descent(self.accounts.dcg_program, handle.record, executor, book.document[0],
                                            round_.descendants, round_.tree_root, fixpoint_accounts=fixpoint))
            result.append(ix.descend(self.accounts.dcg_program, handle.record, challenger, book.document[0],
                                     round_.choice, fixpoint_accounts=fixpoint, document_writable=round_.fixpoint))
        return tuple(result)

    def follow_descent(self, handle: ChallengeHandle, executor: Pubkey | str | bytes | bytearray,
                       challenger: Pubkey | str | bytes | bytearray,
                       rounds: Sequence[DescentRound]) -> tuple[Instruction, ...]:
        return self.descent_instructions(handle, executor, challenger, rounds)

    def read_challenge(self, record: Pubkey | str | bytes | bytearray) -> dict:
        account = self.rpc.read_account(record)
        if account is None:
            raise LookupError("challenge account is absent")
        return c.decode_challenge_record(account.data)

    def ruling(self, record: Pubkey | str | bytes | bytearray) -> dict:
        state = self.read_challenge(record)
        if state["phase"] != c.PHASE_RULED:
            raise RuntimeError("challenge is not ruled")
        return state

    def events(self, signature: str) -> list[dict]:
        return self.rpc.events_from_logs(signature, program=self.accounts.dcg_program)

    def send(self, instruction: Instruction, payer: Keypair, signers: Sequence[Keypair] = ()) -> str:
        receipt = self.rpc.send_transaction([instruction], payer, signers)
        return receipt.signature

File: sdk/typescript/src/watcher.ts
import { Keypair, PublicKey, TransactionInstruction } from "@solana/web3.js";
import type { PathLike } from "node:fs";
import * as c from "./consensus.js";
import * as ix from "./instructions.js";
import { RpcClient } from "./transport.js";

export type KeyLike = PublicKey | string | Uint8Array;

export interface WatcherAccountsOptions {
  pt2s: KeyLike;
  pt1s: KeyLike;
  routes: KeyLike;
  geometry: KeyLike;
  registry: KeyLike;
  pt2sSha256?: Uint8Array;
  pt2s_sha256?: Uint8Array;
}

export class WatcherAccounts {
  readonly dcgProgram: PublicKey;
  readonly pt2s: PublicKey;
  readonly pt1s: PublicKey;
  readonly routes: PublicKey;
  readonly geometry: PublicKey;
  readonly registry: PublicKey;
  readonly pt2sSha256?: Buffer;

  private constructor(dcgProgram: PublicKey, pt2s: PublicKey, pt1s: PublicKey, routes: PublicKey, geometry: PublicKey, registry: PublicKey, pt2sSha256?: Buffer) {
    this.dcgProgram = dcgProgram;
    this.pt2s = pt2s;
    this.pt1s = pt1s;
    this.routes = routes;
    this.geometry = geometry;
    this.registry = registry;
    this.pt2sSha256 = pt2sSha256;
  }

  static create(dcgProgram: KeyLike, options: WatcherAccountsOptions): WatcherAccounts {
    const hash = options.pt2sSha256 ?? options.pt2s_sha256;
    return new WatcherAccounts(c.key(dcgProgram), c.key(options.pt2s), c.key(options.pt1s), c.key(options.routes), c.key(options.geometry), c.key(options.registry), hash === undefined ? undefined : c.buffer(hash));
  }

  get dcg_program(): PublicKey { return this.dcgProgram; }
  get pt2s_sha256(): Buffer | undefined { return this.pt2sSha256; }
}

export class DescentRound {
  readonly descendants: readonly Uint8Array[];
  readonly treeRoot?: Uint8Array;
  readonly choice: number;
  readonly fixpoint: boolean;

  constructor(descendants: readonly Uint8Array[], treeRoot: Uint8Array | undefined, choice: number, fixpoint = false) {
    this.descendants = descendants;
    this.treeRoot = treeRoot;
    this.choice = choice;
    this.fixpoint = fixpoint;
  }
}

export interface ChallengeHandle {
  descriptor: Buffer;
  record: Uint8Array;
  nonce: number;
  position: number;
  instruction: TransactionInstruction;
  response: Uint8Array;
}

export class Watcher {
  readonly rpc: RpcClient;
  readonly accounts: WatcherAccounts;

  constructor(rpc: RpcClient, dcgProgram: KeyLike, accounts: WatcherAccounts) {
    this.rpc = rpc;
    this.accounts = accounts;
    if (c.key(dcgProgram).equals(accounts.dcgProgram) === false) throw new Error("watcher program does not match account map");
  }

  book(descriptor: Uint8Array, options: { challenger?: KeyLike; nonce?: number; positionCount?: number } = {}): c.AddressBook {
    return c.addresses(this.accounts.dcgProgram, { descriptor, pt2s: this.accounts.pt2s.toBytes(), pt2sSha256: this.accounts.pt2sSha256, challenger: options.challenger, nonce: options.nonce ?? 0, registry: this.accounts.registry.toBytes(), positionCount: options.positionCount });
  }

  async readDocument(descriptor: Uint8Array): Promise<c.Dcm2V5> {
    const value = requiredPda(this.book(descriptor).document, "document");
    const account = await this.rpc.readAccount(value[0]);
    if (account === null) throw new Error("document account is absent");
    return c.Dcm2V5.decode(account.data);
  }

  async read_document(descriptor: Uint8Array): Promise<c.Dcm2V5> { return this.readDocument(descriptor); }

  async positions(descriptor: Uint8Array): Promise<Buffer[]> {
    const value = requiredPda(this.book(descriptor).positions, "positions");
    const account = await this.rpc.readAccount(value[0]);
    if (account === null) throw new Error("positions account is absent");
    return c.decodeDpr2(account.data, { allowPartial: true }).roots;
  }

  async listPositions(descriptor: Uint8Array): Promise<Array<[number, Buffer]>> {
    return (await this.positions(descriptor)).map((root, index) => [index, root]);
  }

  async list_positions(descriptor: Uint8Array): Promise<Array<[number, Buffer]>> { return this.listPositions(descriptor); }

  pickSample(descriptor: Uint8Array, positionCount: number, nonce = 0): number {
    if (positionCount < 1 || nonce < 0 || nonce >= 2 ** 64) throw new Error("invalid position count or nonce");
    const value = c.sha256(c.digest(descriptor), c.uint(nonce, 8));
    return Number(value.readBigUInt64LE(0) % BigInt(positionCount));
  }

  pick_sample(descriptor: Uint8Array, positionCount: number, nonce = 0): number { return this.pickSample(descriptor, positionCount, nonce); }

  openPosition(descriptor: Uint8Array, challenger: KeyLike, position: number, options: { responseLength: number; nonce: number }): ChallengeHandle {
    const who = c.keyBytes(challenger);
    const book = this.book(descriptor, { challenger: who, nonce: options.nonce });
    const record = requiredPda(book.challenge, "challenge");
    const data = c.encodeChallengePosition(descriptor, position, options.responseLength, options.nonce);
    const instruction = ix.challengePosition(this.accounts.dcgProgram, record[0], who, requiredPda(book.document, "document")[0], requiredPda(book.positions, "positions")[0], this.accounts.pt2s, this.accounts.routes, this.accounts.geometry, this.accounts.registry, descriptor, position, options.responseLength, options.nonce);
    return { descriptor: c.buffer(descriptor), record: record[0].toBytes(), nonce: options.nonce, position, instruction, response: requiredPda(book.response, "response")[0].toBytes() };
  }

  open_position(descriptor: Uint8Array, challenger: KeyLike, position: number, options: { response_length: number; nonce: number }): ChallengeHandle {
    return this.openPosition(descriptor, challenger, position, { responseLength: options.response_length, nonce: options.nonce });
  }

  openLeaf(descriptor: Uint8Array, challenger: KeyLike, options: { position: number; segment: number; local: number; leaf: Uint8Array; path: readonly Uint8Array[]; spp1: Uint8Array; responseLength: number; nonce: number }): ChallengeHandle {
    const who = c.keyBytes(challenger);
    const book = this.book(descriptor, { challenger: who, nonce: options.nonce });
    const record = requiredPda(book.challenge, "challenge");
    const data = c.encodeChallengeLeaf(descriptor, options.position, options.segment, options.local, options.leaf, options.path, options.spp1, options.responseLength, options.nonce);
    const instruction = ix.challengeLeaf(this.accounts.dcgProgram, record[0], who, requiredPda(book.document, "document")[0], requiredPda(book.positions, "positions")[0], this.accounts.pt2s, this.accounts.routes, this.accounts.geometry, this.accounts.registry, this.accounts.pt1s, data);
    return { descriptor: c.buffer(descriptor), record: record[0].toBytes(), nonce: options.nonce, position: options.position, instruction, response: requiredPda(book.response, "response")[0].toBytes() };
  }

  open_leaf(descriptor: Uint8Array, challenger: KeyLike, options: { position: number; segment: number; local: number; leaf: Uint8Array; path: readonly Uint8Array[]; spp1: Uint8Array; response_length: number; nonce: number }): ChallengeHandle {
    return this.openLeaf(descriptor, challenger, { ...options, responseLength: options.response_length });
  }

  revealPosition(handle: ChallengeHandle, executor: KeyLike, first: number, roots: readonly Uint8Array[]): TransactionInstruction {
    const book = this.book(handle.descriptor);
    return ix.revealPosition(this.accounts.dcgProgram, handle.record, executor, requiredPda(book.document, "document")[0], requiredPda(book.positions, "positions")[0], this.accounts.pt2s, this.accounts.routes, this.accounts.geometry, first, roots);
  }

  reveal_position(handle: ChallengeHandle, executor: KeyLike, first: number, roots: readonly Uint8Array[]): TransactionInstruction { return this.revealPosition(handle, executor, first, roots); }

  selectSegment(handle: ChallengeHandle, challenger: KeyLike, ordinal: number): TransactionInstruction {
    const book = this.book(handle.descriptor);
    return ix.selectSegment(this.accounts.dcgProgram, handle.record, challenger, requiredPda(book.document, "document")[0], this.accounts.pt2s, this.accounts.routes, this.accounts.geometry, ordinal);
  }

  select_segment(handle: ChallengeHandle, challenger: KeyLike, ordinal: number): TransactionInstruction { return this.selectSegment(handle, challenger, ordinal); }

  closeResponse(handle: ChallengeHandle, executor: KeyLike): TransactionInstruction {
    const response = handle.response.length === 0 ? c.pda(this.accounts.dcgProgram, c.RESPONSE_SEED, handle.record)[0] : new PublicKey(handle.response);
    return ix.closeResponse(this.accounts.dcgProgram, handle.record, response, executor);
  }

  close_response(handle: ChallengeHandle, executor: KeyLike): TransactionInstruction { return this.closeResponse(handle, executor); }

  async settle(handle: ChallengeHandle, winner: KeyLike, challenger: KeyLike, executor?: KeyLike): Promise<TransactionInstruction> {
    const response = handle.response.length === 0 ? c.pda(this.accounts.dcgProgram, c.RESPONSE_SEED, handle.record)[0] : new PublicKey(handle.response);
    const resolvedExecutor = executor ?? (await this.readChallenge(handle.record)).executor;
    const book = this.book(handle.descriptor);
    return ix.settle(this.accounts.dcgProgram, handle.record, response, winner, resolvedExecutor, requiredPda(book.document, "document")[0], challenger);
  }

  private fixpointAccounts(): ix.AccountSpec[] {
    return [[this.accounts.pt2s, false, false], [this.accounts.routes, false, false], [this.accounts.geometry, false, false], [this.accounts.registry, false, false], [this.accounts.pt1s, false, false]];
  }

  descentInstructions(handle: ChallengeHandle, executor: KeyLike, challenger: KeyLike, rounds: readonly DescentRound[]): TransactionInstruction[] {
    const book = this.book(handle.descriptor);
    const result: TransactionInstruction[] = [];
    for (const round of rounds) {
      const fixpoint = round.fixpoint ? this.fixpointAccounts() : undefined;
      result.push(ix.revealDescent(this.accounts.dcgProgram, handle.record, executor, requiredPda(book.document, "document")[0], round.descendants, round.treeRoot, { fixpointAccounts: fixpoint }));
      result.push(ix.descend(this.accounts.dcgProgram, handle.record, challenger, requiredPda(book.document, "document")[0], round.choice, { fixpointAccounts: fixpoint, documentWritable: round.fixpoint }));
    }
    return result;
  }

  descent_instructions(handle: ChallengeHandle, executor: KeyLike, challenger: KeyLike, rounds: readonly DescentRound[]): TransactionInstruction[] { return this.descentInstructions(handle, executor, challenger, rounds); }

  followDescent(handle: ChallengeHandle, executor: KeyLike, challenger: KeyLike, rounds: readonly DescentRound[]): TransactionInstruction[] { return this.descentInstructions(handle, executor, challenger, rounds); }
  follow_descent(handle: ChallengeHandle, executor: KeyLike, challenger: KeyLike, rounds: readonly DescentRound[]): TransactionInstruction[] { return this.followDescent(handle, executor, challenger, rounds); }

  async readChallenge(record: KeyLike): Promise<ReturnType<typeof c.decodeChallengeRecord>> {
    const account = await this.rpc.readAccount(c.key(record));
    if (account === null) throw new Error("challenge account is absent");
    return c.decodeChallengeRecord(account.data);
  }

  async read_challenge(record: KeyLike): Promise<ReturnType<typeof c.decodeChallengeRecord>> { return this.readChallenge(record); }

  async ruling(record: KeyLike): Promise<ReturnType<typeof c.decodeChallengeRecord>> {
    const state = await this.readChallenge(record);
    if (state.phase !== c.PHASE_RULED) throw new Error("challenge is not ruled");
    return state;
  }

  async events(signature: string): Promise<Array<Record<string, c.Integer | Buffer | string>>> { return this.rpc.eventsFromLogs(signature, this.accounts.dcgProgram); }

  async send(instruction: TransactionInstruction, payer: Keypair | string | PathLike, signers: readonly (Keypair | string | PathLike)[] = []): Promise<string> {
    const receipt = await this.rpc.sendTransaction([instruction], payer, signers);
    return receipt.signature;
  }
}

function requiredPda(value: c.Pda | undefined, name: string): c.Pda {
  if (value === undefined) throw new Error(`missing ${name} address`);
  return value;
}
